### PR TITLE
Implement layout_minWidth and layout_minHeight

### PR DESCRIPTION
**Pull request by evant**
This allows you to specify a minimum size your view is allowed to
shrink to. #1

I plan to implement max-width/height as well in a separate commit, but I want to make sure I'm on the right track first.

### CODE REVIEW

**Comment by thagikura** [comment]
Wow thank you! I won't have access to my laptop for a day or so, once I have one, let me start reviewing your code, thanks!

File: app/src/androidTest/java/com/google/android/apps/flexbox/test/MainActivityTest.java [hunk 1]
```diff
@@ -16,19 +16,6 @@
 
 package com.google.android.apps.flexbox.test;
 
-import static android.support.test.espresso.Espresso.onView;
```
Comment (thagikura) [inline app/src/androidTest/java/com/google/android/apps/flexbox/test/MainActivityTest.java]: Please restore the import order here.
I've updated the CONTRIBUTING.md to follow the official AndroidStyle code style.
Sorry wasn't updated before this PR.

File: flexbox/src/androidTest/java/com/google/android/flexbox/test/FlexboxAndroidTest.java [hunk 2]
```diff
@@ -17,14 +17,10 @@
 package com.google.android.flexbox.test;
 
 
-import com.google.android.flexbox.FlexboxLayout;
```
Comment (thagikura) [inline flexbox/src/androidTest/java/com/google/android/flexbox/test/FlexboxAndroidTest.java]: Ditto

File: flexbox/src/main/res/layout/activity_minwidth_test.xml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (thagikura) [inline flexbox/src/main/res/layout/activity_minwidth_test.xml]: Please put this file under flexbox/src/androidTest/res
This file is used only for testing.

File: flexbox/src/main/res/layout/activity_minwidth_test.xml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+<com.google.android.flexbox.FlexboxLayout android:id="@+id/flexbox_layout"
```
Comment (thagikura) [inline flexbox/src/main/res/layout/activity_minwidth_test.xml]: Also please add license notice copied from other layout files.

File: flexbox/src/main/res/layout/activity_minheight_test.xml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (thagikura) [inline flexbox/src/main/res/layout/activity_minheight_test.xml]: Please put this file under flexbox/src/androidTest/res
and put license notice here

File: flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java [hunk 6]
```diff
@@ -614,64 +623,91 @@ private void determineMainSize(@FlexDirection int flexDirection, int widthMeasur
      */
     private int expandFlexItems(FlexLine flexLine, @FlexDirection int flexDirection,
             int maxMainSize, int paddingAlongMainAxis, int startIndex) {
+        int childIndex = startIndex;
```
Comment (thagikura) [inline flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java]: expandFlexItems doesn't need to be changed unless layout_maxHeight/layout_maxWidth are implemented.
Could you revert the change for the expandFlexItems?

File: flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java [hunk 7]
```diff
@@ -212,6 +214,12 @@
 
     private List<FlexLine> mFlexLines = new ArrayList<>();
 
+    /**
+     * Holds the 'frozen' state of children during measure. If a view is frozen it will no longer
+     * expand or shrink regardless of flexGrow/flexShrink.
```
Comment (thagikura) [inline flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java]: Could you add comment that this field refers to the reordered indices?

File: flexbox/src/main/res/values/attrs.xml [hunk 8]
```diff
@@ -99,5 +99,8 @@ limitations under the License.
             <enum name="baseline" value="3" />
             <enum name="stretch" value="4" />
         </attr>
+
+        <attr name="layout_minWidth" format="dimension" />
```
Comment (thagikura) [inline flexbox/src/main/res/values/attrs.xml]: Also could you update the FlexItem class under the app module?
Otherwise layout_minWidth/layout_minHeight aren't restored after a configuration change for the demo app.

File: flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java [hunk 9]
```diff
@@ -715,59 +751,67 @@ private int shrinkFlexItems(FlexLine flexLine, @FlexDirection int flexDirection,
...
-                    newWidth = 0;
-                } else {
-                    accumulatedRoundError += (rawCalculatedWidth - newWidth);
-                    if (accumulatedRoundError > 1.0) {
-                        newWidth += 1;
-                        accumulatedRoundError -= 1;
-                    } else if (accumulatedRoundError < -1.0) {
-                        newWidth -= 1;
-                        accumulatedRoundError += 1;
+                if (!mFrozenChildren[childIndex]) {
```
Comment (thagikura) [inline flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java]: I think this condition check can be at the line 752 before the if statement:
```java
if (flexDirection == FLEX_DIRECTION_ROW
                      || flexDirection == FLEX_DIRECTION_ROW_REVERSE) {		                      || flexDirection == FLEX_DIRECTION_ROW_REVERSE)
```

like 
```java
if (mFrozenChildren[childIndex]) {
    childIndex++;
    continue;
}
```

To avoid the duplicate if checks in the line 754 and 788 and to reduce the indent

File: flexbox/src/main/res/layout/activity_minheight_test.xml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    android:layout_height="match_parent"
+    app:flexDirection="column">
+
+    <TextView
+        android:id="@+id/text1"
+        android:layout_width="wrap_content"
+        android:layout_height="0dp"
+        android:text="1"
+        app:layout_flexGrow="1"
+        app:layout_minHeight="300dp" />
```
Comment (thagikura) [inline flexbox/src/main/res/layout/activity_minheight_test.xml]: I think this test case can't be always true.
In fact it failed with my environment (Nexus5x emulator)

In some situations, text1 isn't trying shrink to less than 300dp
(e.g. If the parent's height is 800dp, then text1 and text2 are trying to be expanded to 400dp.
Could you make the test deterministic?
Also the same for the minwidth test.

**Comment by thagikura** [comment]
Took a first round of look. Could you confirm the comments? Thanks!

File: flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java (hunk 6 above)
Comment (evant) [inline flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java]: As per https://www.w3.org/TR/css-flexbox-1/#resolve-flexible-lengths min constraints are only checked after space is distributed. This could mean that an item was expanded but did not meet the min size. (Say it had a width of wrap content and happened to be less than the min width and it did not receive enough free space when expanding).

File: flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java (hunk 9 above)
Comment (evant) [inline flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java]: Regardless of if the child is frozen or not it still needs to run
```java
flexLine.mainSize += child.getMeasuredWidth() + lp.leftMargin + lp.rightMargin;
```
or
```java
flexLine.mainSize += child.getMeasuredHeight() + lp.topMargin + lp.bottomMargin;
```
for the main size to be correct. I'm all for reducing intention but I don't see a clean way to do this without duplicating those lines.

File: flexbox/src/main/res/values/attrs.xml (hunk 8 above)
Comment (evant) [inline flexbox/src/main/res/values/attrs.xml]: Updated sample app to support min-width and min-height. I chose to make this a separate commit, let me know if you think it would be batter squashed.

**Comment by evant** [comment]
Updated in response to above comments

File: app/src/main/java/com/google/android/apps/flexbox/FlexItem.java [hunk 11]
```diff
@@ -124,6 +132,8 @@ protected FlexItem(Parcel in) {
         MarginLayoutParamsCompat.setMarginStart(lp, startMargin);
         MarginLayoutParamsCompat.setMarginEnd(lp, endMargin);
         lp.bottomMargin = bottomMargin;
+        lp.minWidth = minWidth;
```
Comment (thagikura) [inline app/src/main/java/com/google/android/apps/flexbox/FlexItem.java]: For the width/height fields, values are stored as DP in the FlexItem class and converting to LayoutParams, convert it as pixels.

Could you follow that way for also minWidth and minHeight?

i.e. 
lp.minWidth = Util.dpToPixel(context, minWidth)
lp.minHeight = Util.dpToPixel(context, minHeight)

File: app/src/main/java/com/google/android/apps/flexbox/FlexItem.java [hunk 12]
```diff
@@ -146,6 +156,8 @@ public static FlexItem fromFlexView(View view, int index) {
         flexItem.paddingStart = ViewCompat.getPaddingStart(view);
         flexItem.paddingEnd = ViewCompat.getPaddingEnd(view);
         flexItem.paddingBottom = view.getPaddingBottom();
+        flexItem.minWidth = lp.minWidth;
```
Comment (thagikura) [inline app/src/main/java/com/google/android/apps/flexbox/FlexItem.java]: Also could you follow the width/height fields as 

flexItem.minWidth = Util.pixelToDp(view.getContext, lp.minWidth);
flexItem.minHeight = Util.pixelToDp(view.getContext, lp.minHeight);

File: flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java (hunk 6 above)
Comment (thagikura) [inline flexbox/src/main/java/com/google/android/flexbox/FlexboxLayout.java]: To check the min constraints, I think adding the following check in the first loop in the measureHorizontal/measureVertical is more simple and readable.
(after the child.measure(childWidthMeasureSpec, childHeightMeasureSpec); in line 371 for measureHorizontal 
and line 519 for measureVertical)

In the expandFlexItems, there is no chance that the size of the flex items are reduced, so min constraints violation can't be happen in expandFlexItems.
So I think only checking max constraints violation in expandFlexItems is more simple.

```java
 if (child.getMeasuredWidth() < lp.minWidth) {
                    child.measure(MeasureSpec.makeMeasureSpec(lp.minWidth, MeasureSpec.EXACTLY),
                            MeasureSpec.makeMeasureSpec(child.getMeasuredHeight(),
                                    MeasureSpec.EXACTLY));
                }
```

or for measureVertical

```java
 if (child.getMeasuredHeight() < lp.minHeight) {
                    child.measure(MeasureSpec.makeMeasureSpec(child.getMeasuredWidth(), MeasureSpec.EXACTLY),
                            MeasureSpec.makeMeasureSpec(lp.minHeight,
                                    MeasureSpec.EXACTLY));
                }
```

File: flexbox/src/main/res/values/attrs.xml (hunk 8 above)
Comment (thagikura) [inline flexbox/src/main/res/values/attrs.xml]: The separate commit for the demo app sounds good, thanks.

File: flexbox/src/androidTest/res/layout/activity_minheight_test.xml [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        android:text="1"
+        app:layout_flexGrow="1"
+        app:layout_minHeight="300dp" />
+
+    <TextView
+        android:id="@+id/text2"
+        android:layout_width="wrap_content"
+        android:layout_height="0dp"
+        android:text="2"
+        app:layout_flexGrow="1" />
```
Comment (thagikura) [inline flexbox/src/androidTest/res/layout/activity_minheight_test.xml]: How about using a fixed layout_height for the FlexboxLayout?
This test isn't still deterministic. It still failed in my Nexus5x device.

E.g. if the parent's height is 800dp, then text1 and text2 will be expanded to 400dp.

File: flexbox/src/androidTest/res/layout/activity_minwidth_test.xml [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  -->
+<com.google.android.flexbox.FlexboxLayout android:id="@+id/flexbox_layout"
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
```
Comment (thagikura) [inline flexbox/src/androidTest/res/layout/activity_minwidth_test.xml]: Also how about using a fixed value for the parent's width?
It succeeded with my Nexus5x, but I think it's not deterministic in some devices.

Comment (thagikura) [inline flexbox/src/androidTest/res/layout/activity_minwidth_test.xml]: Thanks. Had a second round of look and left some comments.

**Comment by thagikura** [comment]
Hi, thanks for your continuous effort.
If any of my review comments are not clear, please feel free to let me know.

**Comment by thagikura** [comment]
Hi, thank you for your contribution!

I self addressed my review comments in another PR based on your PR.
https://github.com/google/flexbox-layout/pull/51

So let me close this PR.
